lib/ts/MemArena.h

```diff
@@ -0,0 +1,159 @@
+      @return a MemSpan of the allocated memory.
+   */
+  MemSpan alloc(size_t n);
+
+  /** MemArena& freeze(size_t n = 0)
+
+      Will "freeze" a generation of memory. Any memory previously allocated can still be used. This is an
+      important distinction as freeze does not mean that the memory is immutable, only that subsequent allocations
+      will be in a new generation.
+
+      @param n Number of bytes for new generation.
+        if @a n == 0, the next generation will be large enough to hold all existing allocations.
+      @return @c *this
+   */
+  MemArena &freeze(size_t n = 0);
+
+  /** MemArena& thaw()
+
+      Will "thaw" away any previously frozen generations. Any generation that is not the current generation is considered
+      frozen because there is no way to allocate in any of those memory blocks. thaw() is the only mechanism for deallocating
+      memory in the arena (other than destroying the arena itself). Thawing away previous generations means that all spans
+      of memory allocated in those generations are no longer safe to use.
+
+      @return @c *this
+   */
+  MemArena &thaw();
+
+  /** MemArena& empty
+
+      Empties the entire arena and deallocates all underlying memory. Next block size will be equal to the sum of all
+      allocations before the call to empty.
+   */
+  MemArena &empty();
+
+  /// @returns the current generation @c size.
+  size_t
+  size() const
+  {
+    return arena_size;
+  }
+
+  /// @returns the @c remaining space within the generation.
+  size_t
+  remaining() const
+  {
+    return (current) ? current->size - current->allocated : 0;
+  }
+
+  /// @returns the total number of bytes allocated within the arena.
+  size_t
+  allocated_size() const
+  {
+    return total_alloc;
+  }
+
+  /// @returns the number of bytes that have not been allocated within the arena
+  size_t
+  unallocated_size() const
+  {
+    return size() - allocated_size();
+  }
+
+  /// @return a @c true if @ptr is in memory owned by this arena, @c false if not.
+  bool contains(void *ptr) const;
+
+private:
+  /// creates a new @Block of size @n and places it within the @allocations list.
+  /// @return a pointer to the block to allocate from.
+  Block *newInternalBlock(size_t n, bool back);
+
+  static constexpr size_t DEFAULT_BLOCK_SIZE = 1 << 15; ///< 32kb
+  static constexpr size_t DEFAULT_PAGE_SIZE  = 1 << 12; ///< 4kb
+  /** generation_size and prev_alloc are used to help quickly figure out the arena
+        info (arena_size and total_alloc) after a thaw().
+   */
+  size_t arena_size      = DEFAULT_BLOCK_SIZE; ///< --all
+  size_t generation_size = 0;                  ///< Size of current generation -- all
+  size_t total_alloc     = 0;                  ///< Total number of bytes allocated in the arena -- allocated
+  size_t prev_alloc      = 0;                  ///< Total allocations before current generation -- allocated
+
+  bool next_generation   = false;
```

Comment:
Why have this, instead of just checking if `generation` is `nullptr`?

Reply:
good point. changed.